_posts/2018-08-01-gke-multi-stage.md

```diff
@@ -0,0 +1,525 @@
+      servicePort: 8080
+      path: /
+  tls:
+    - secretName: openfaas-cert
+      hosts:
+      - openfaas.example.com
+affinity:
+  nodeAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: cloud.google.com/gke-preemptible
+          operator: DoesNotExist
+  podAntiAffinity:
+    preferredDuringSchedulingIgnoredDuringExecution:
+      - weight: 100
+        podAffinityTerm:
+          labelSelector:
+            matchLabels:
+              app: gateway
+              release: openfaas-prod
+          topologyKey: kubernetes.io/hostname
+```
+
+For production the OpenFaaS gateway is scaled to two replicas and with the pod anti-affinity rule we make sure that each replica will run on a different node. 
+Note that `operator.createCRD` is set to false since the `functions.openfaas.com` custom resource definition is already present on the cluster.
+
+Save the above file as `openfaas-prod.yaml` and install OpenFaaS instance from the project helm repository:
+
+```bash
+helm upgrade openfaas-prod --install openfaas/openfaas \
+    --namespace openfaas-prod  \
+    -f openfaas-prod.yaml
+```
+
+## OpenFaaS functions operations
+
+![openfaas-operator](/images/gke-multi-stage/operator.png)
+
+Using the OpenFaaS CRD you can define functions as Kubernetes custom resource:
+
+```yaml
+apiVersion: openfaas.com/v1alpha2
+kind: Function
+metadata:
+  name: certinfo
+spec:
+  name: certinfo
+  image: stefanprodan/certinfo:latest
+  labels:
+    com.openfaas.scale.min: "2"
+    com.openfaas.scale.max: "12"
+    com.openfaas.scale.factor: "4"
+  limits:
+    cpu: "1000m"
+    memory: "128Mi"
+  requests:
+    cpu: "10m"
+    memory: "64Mi"
+  constraints:
+    - "cloud.google.com/gke-preemptible=true"
+```
+
+Note that this function will be running on the fn pool due to the affinity constraint `cloud.google.com/gke-preemptible=true`.
+
+Save the above resource as `certinfo.yaml` and use `kubectl` to deploy the function on both instances:
+
+```bash
+kubectl -n openfaas-stg-fn apply -f certinfo.yaml
+kubectl -n openfaas-prod-fn apply -f certinfo.yaml
+```
+
+Verify both endpoints are live:
+
+```
+curl -d "openfaas-stg.example.com" https://openfaas-stg.example.com/function/certinfo
+
+Issuer Let's Encrypt Authority X3
+....
+
+curl -d "openfaas.example.com" https://openfaas.example.com/function/certinfo
+
+Issuer Let's Encrypt Authority X3
+....
+```
+
+To manage functions on Kubernetes you can use kubectl but for development you'll need the OpenFaaS CLI. 
+You can install faas-cli and connect to the staging instance with:
+
+```bash
+curl -sL https://cli.openfaas.com | sudo sh
+
+echo $stg_password | faas-cli login -u admin --password-stdin -g https://openfaas-stg.example.com
+```
+
+Using faas-cli and kubectl a development workflow would look like this:
+
+* Create a function from a code template `faas-cli new myfn --lang go --prefix gcr.io/project-id` 
```

Comment:
Formatting of this section looks a bit cluttered, don't have an alternative to suggest tho.